### ISSUE TITLE
single pixel outlier

### ISSUE THREAD

**Issue by JustineJiao** [MouseLand/cellpose#516]
Hi there,

When I segment polyonal cells using Cellpose, I found some single pixel outlier. 
Like in this labeled image, that pixel has the same label ID that the brown big cell to the right. When I processed this labeled image in LabelsToROIs plugin (https://github.com/ariel-waisman/LabelsToROIs), this single pixel would be recognized as one cell, which will affect cell size calculation.

We suspect this is a glitch of Cellpose. Generally, it's not a big problem, but I think you should know.

Does anyone faced this situation?

Justine Jiao

![image](https://user-images.githubusercontent.com/58281975/169438218-21ccf4d6-6389-4268-81c5-5a371b077c5b.png)

![image](https://user-images.githubusercontent.com/58281975/169440926-8a4cdebf-1c6a-4090-8a2d-6985e85a8fad.png)

**Comment by mayishazn** [comment]
On the jupyter notebook version I have a line in my code setting a minarea=100 and then
masks, flows, styles = model.eval(imgs, diameter=diam, channels=channels, do_3D=False, min_size=minarea, resample=True, progress=True)

It takes care of impossibly small cells for me. Hope this helps!

**Comment by JustineJiao** [comment]
@mayishazn  I tried this method but  it didn't work. The single pixel belongs to the big label to the right, which means that they are one large label, so the single pixel can not filter by its size.
Thanks anyway.

**Comment by LeahdamF** [comment]
This is exactly the issue I'm having: single pixels belonging to a neighboring big cell. Has anyone found a solution?

**Comment by lopollar** [comment]
When segmenting nuclei, I often end up with similar issues, where one or two pixels end up outside of the shape. However, when performing segmentation on whole cells, I often get more pixels not being connected (more often=10% of the segmentation masks aren't connected). 
When performing segmentation and then looking at the segmentation mask of one cell (in this case cell 3), I end up with this shape:
![image](https://user-images.githubusercontent.com/52383997/222143022-90e37712-715c-4f2e-9e06-d309281fa862.png)
Is there any explanation on how this is possible? And how to solve it? 
This greatly influences the downstream analysis. 
I can give way more context, examples and example code if wanted. 
In general though I am extremely happy with the results of cellpose.
Thanks a lot,

Lotte

**Comment by carsen-stringer** [comment]
thanks @lopollar , can you give more context for your problem? is that in 3D or in 2D segmentation? your case seems a bit different

we can add an option to run a connected components check after the segmentation to remove single pixel outliers (cc @mrariden )

**Comment by yassinharim** [comment]
As suggested by @mrariden I thought to contribute my experience with the similar issue :)

I'm experiencing an issue with a model to recognise DAPI-stained nuclei in 2D images. If a certain mask is affected, it is correctly generated to mark a nucleus - but then, there is one more pixel, completely detached from the original mask, that Cellpose considers as a part of this mask as well. The issue already happens at the standard settings of `flow_threshold = 0.4` and `cell_prob_threshold = 0.0`, but is pronounced at settings of `flow_threshold = 0.6` and `cell_prob_threshold = -1.0` that generally improve detection of cells in the model. 
This causes issues in my downstream analysis, as I am importing ROIs into ImageJ from label images and this situation would then generate two separate ROIs from one gray value in the label image, throwing off the ROI counting. 

This issue seems to be restricted to several versions (including the most current ones) of the model - earlier versions do not have this problem.

I'm attaching some screenshots below for you, as well as attaching the model and a sample image (ZIP because the file types were not allowed). 

![Bug1](https://github.com/MouseLand/cellpose/assets/127263444/b39274f4-c2d3-4962-9a4d-61917753dbda)
![Bug2](https://github.com/MouseLand/cellpose/assets/127263444/6fe10cd8-0779-4c10-b548-b82c937037b9)

Here's another example of a mask from a model with `flow_threshold = 0.9` and `cell_prob_threshold = -3.0` that is basically a line, and is also split in two different parts:

![One-dimensional mask](https://github.com/MouseLand/cellpose/assets/127263444/d8b39523-66ad-4877-b815-9e8cd2c15042)

[DAPI_40X.zip](https://github.com/MouseLand/cellpose/files/11870893/DAPI_40X.zip)
[Sample image.zip](https://github.com/MouseLand/cellpose/files/11870903/Sample.image.zip)

**Comment by yassinharim** [comment]
Hey, I just wanted to bring up this issue again and ask if there are any plans to implement this post-processing for connected components in the future?

**Comment by carsen-stringer** [comment]
we can add it as an optional post-processing step, a PR is welcome
